keps/0009-ingress-class-params-crd.md

```diff
@@ -0,0 +1,64 @@
+---
+title: Introduce IngressClassParams CRD to control IngressClass behavior
+status: provisional
+---
+
+# Introduce IngressClassParams CRD to control IngressClass behavior
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+As of now users can control the behavior of an `IngressClass` via e.g.
+[`ingress.kubernetes.io/service-upstream` annotation][service-upstream-annotation].
+This is handy in general but gets cumbersome when one needs to change this for
+multiple services.
+
+Controlling `IngressClass` parameters in one place would be much more desirable in
+such situations hence the proposal to introduce a new object - `IngressClassParams` -
+which would allow said customizations.
+
+[service-upstream-annotation]: https://docs.konghq.com/kubernetes-ingress-controller/2.3.x/references/annotations/#ingresskubernetesioservice-upstream
+
+## Motivation
+
+- we want allow `IngressClass` behavior customizations in one place
+- we want to make configuration management related to `IngressClass` to not be
+  a burden for the user even when IngressClass is assigned to many services
+
+### Goals
+
+- provide a definition of new object - `IngressClassParams`
+- provide an _ability to configure_ `IngressClass` in one place via aforementioned
+  `IngressClassParams` object
+- provide `ServiceUpstream` field in `IngressClassParams` object, as means to control
+  the behavior already provided by `ingress.kubernetes.io/service-upstream` annotation
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### Graduation Criteria
+
+- [ ] `IngressClassParams` object available for deployment for end users
+- [ ] `ServiceUpstream` field implemented in `IngressClassParams` object to allow
+  control of behavior of all services handled by an `IngressClass`
```

Comment:
I think we should be modeling our graduation criteria around an initial period of time where the API is locked behind a feature gate as we've defined [here](https://github.com/Kong/kubernetes-ingress-controller/blob/main/FEATURE_GATES.md) and starts at a `v1alpha1`. Doing so gives us the freedom to:

1. make backwards incompatible changes between minor releases
2. use telemetry to see if end-users are using the feature, without being committed to the feature
3. potentially completely remove the feature and reducing our maintenance burden if we don't see usage

One of the main reasons why I feel this way is because of some of the context in:

- https://github.com/Kong/kubernetes-ingress-controller/issues/1131
- https://github.com/Kong/kubernetes-ingress-controller/issues/774

Which this API originates from. Consider these assertions based on the history of this feature request:

1. the request to make this happen has _only_ ever been initiated or substantiated by internal Kong team members, **no end-users**
2. the only end-user engagement we've got about this feature, was a prominent community member basically suggesting that we _don't_ do it

With that in mind we should consider whether we really want to do this at all. And if we do, let's make it easy on ourselves to make changes or even quickly back out of it entirely. By re-framing this graduation criteria starting with an initial soaking period behind a feature gate, and as a `v1alpha1`.

Reply:
That's a good idea. How should we approach this though? The logic in the controller can be put behind a feature flag but the CRD would have to go to the manifests anyway, right?

Reply:
I'd like to go ahead and merge this PR but I feel we haven't reached a consensus on this discussion (or maybe I'm lacking context to understand that we've reached it).

> The logic in the controller can be put behind a feature flag but the CRD would have to go to the manifests anyway, right?

This specifically. I'm not sure if putting the CRDs without a feature gate is "OK" or is there a different mechanism that I'm not aware of yet.

Reply:
It's mentioned that we're going to put this feature behind a feature gate, which means we have time to experiment with it before we're committed to it. I don't want to see this KEP getting blocked any longer, and In my mind that's enough for now to get started on an implementation so I consider this resolved.